### ISSUE TITLE
panic: value of type []interface {} is not assignable to type string

### ISSUE THREAD

**Issue by juanjojulian** [hashicorp/terraform#19125]
### Terraform Version

```
2018/10/19 11:14:40 [INFO] Terraform version: 0.11.8  
2018/10/19 11:14:40 [INFO] Go runtime version: go1.10.3
```

### Terraform Configuration Files

There is more code but I think the problem is related to this one:

```hcl
locals {
  unique_regions = ["${distinct(var.region)}"]
  availavilityset_regions     = {
    "${azurerm_availability_set.main.*.location}" = "${azurerm_availability_set.main.*.id}"
  }  
}

resource "azurerm_availability_set" "main" {
  name                     = "${var.sub-name}-${local.unique_regions[count.index]}-${var.aplication}-as"
  location                 = "${local.unique_regions[count.index]}"
  resource_group_name      = "${var.sub-name}-${local.unique_regions[count.index]}-${var.aplication}-rg"
  managed                  = true
  count                    = "${length(local.unique_regions)}"
}

resource "azurerm_virtual_machine" "main" {
  name                  = "${var.hostname[count.index]}"
  location              = "${var.region[count.index]}"
  resource_group_name   = "${var.sub-name}-${var.region[count.index]}-${var.aplication}"
  network_interface_ids = ["${element(azurerm_network_interface.main.*.id, count.index)}"]
  vm_size               = "Standard_B2ms"
  license_type          = "Windows_Server"
  availability_set_id   = "${lookup(local.availavilityset_regions, "${var.region[count.index]}")}"

[...]

```

### Crash Output

```
------------------------------------------------------------------------
panic: reflect.Value.SetMapIndex: value of type []interface {} is not assignable to type string

goroutine 2287 [running]:
reflect.Value.assignTo(0x2ec7b60, 0xc4206faae0, 0x97, 0x39df1f4, 0x19, 0x2f1c280, 0x0, 0x2f1c280, 0xc420848f90, 0x98)
	/usr/local/Cellar/go/1.10.3/libexec/src/reflect/value.go:2235 +0x427
reflect.Value.SetMapIndex(0x325b5e0, 0xc420460148, 0x195, 0x2ec7b60, 0xc4206faae0, 0x97, 0x31fe120, 0xc420848fa0, 0x94)
	/usr/local/Cellar/go/1.10.3/libexec/src/reflect/value.go:1496 +0xb4
github.com/hashicorp/terraform/config.(*interpolationWalker).Primitive(0xc4203ec300, 0x2f1c280, 0xc420848f90, 0x98, 0xc4203ec300, 0x1)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/config/interpolate_walk.go:189 +0x365
github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk.walkPrimitive(0x2f1c280, 0xc420848f90, 0x98, 0x372f060, 0xc4203ec300, 0x100ca00, 0xc42002e000)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk/reflectwalk.go:251 +0x80
github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk.walk(0x2f1c280, 0xc420848f90, 0x98, 0x372f060, 0xc4203ec300, 0xc420848f01, 0x98)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk/reflectwalk.go:179 +0x1ff
github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk.walkMap(0x325b5e0, 0xc420460148, 0x195, 0x372f060, 0xc4203ec300, 0x100ca00, 0xc42002e000)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk/reflectwalk.go:224 +0x28f
github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk.walk(0x325b5e0, 0xc420460148, 0x195, 0x372f060, 0xc4203ec300, 0x1, 0x0)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk/reflectwalk.go:182 +0x3ef
github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk.walkSlice(0x2ec8120, 0xc4206fa9c0, 0x97, 0x372f060, 0xc4203ec300, 0x97, 0x0)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk/reflectwalk.go:283 +0x20d
github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk.walk(0x31fe120, 0xc420848f10, 0x94, 0x372f060, 0xc4203ec300, 0x0, 0x0)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk/reflectwalk.go:185 +0x35c
github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk.walkMap(0x325b5e0, 0xc4206a68d0, 0x15, 0x372f060, 0xc4203ec300, 0x100ca00, 0xc42002e000)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk/reflectwalk.go:233 +0x2f1
github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk.walk(0x325b5e0, 0xc4206a68d0, 0x15, 0x372f060, 0xc4203ec300, 0x1, 0x0)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk/reflectwalk.go:182 +0x3ef
github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk.Walk(0x325b5e0, 0xc4206a68d0, 0x372f060, 0xc4203ec300, 0x0, 0x0)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/reflectwalk/reflectwalk.go:92 +0x131
github.com/hashicorp/terraform/config.(*RawConfig).interpolate(0xc4203d93b0, 0xc420848df0, 0xc4203d93f8, 0x5000105)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/config/raw_config.go:286 +0x14d
github.com/hashicorp/terraform/config.(*RawConfig).Interpolate(0xc4203d93b0, 0xc4208cf560, 0x0, 0x0)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/config/raw_config.go:175 +0xcd
github.com/hashicorp/terraform/terraform.(*BuiltinEvalContext).Interpolate(0xc4200abad0, 0xc4203d93b0, 0x0, 0x2000, 0x3c, 0x16657ecf6)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/terraform/eval_context_builtin.go:249 +0xef
github.com/hashicorp/terraform/terraform.(*EvalLocal).Eval(0xc4206f9260, 0x3b8ce00, 0xc4200abad0, 0x0, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/terraform/eval_local.go:18 +0x76
github.com/hashicorp/terraform/terraform.EvalRaw(0x3b66780, 0xc4206f9260, 0x3b8ce00, 0xc4200abad0, 0x31fc140, 0x57dfe23, 0x2f1c280, 0xc4206ed6e0)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x156
github.com/hashicorp/terraform/terraform.Eval(0x3b66780, 0xc4206f9260, 0x3b8ce00, 0xc4200abad0, 0xc4206f9260, 0x3b66780, 0xc4206f9260, 0xc420082fe0)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/terraform/eval.go:34 +0x4d
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x35636e0, 0xc4206279c0, 0x0, 0x0)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/terraform/graph.go:126 +0xc26
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc42020e9a0, 0x35636e0, 0xc4206279c0, 0xc420719b40)
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/dag/walk.go:387 +0x3a0
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/private/tmp/terraform-20180816-49425-1aa8qy/terraform-0.11.8/src/github.com/hashicorp/terraform/dag/walk.go:310 +0x1248
```

### Expected Behavior

Trying to create code to deploy multiple VM´s in multiple regions with multiple availability sets. VM´s should go the availability set that resides in its region.

```
availavilityset_regions     = {
    "${azurerm_availability_set.main.*.location}" = "${azurerm_availability_set.main.*.id}"
  }
```

Should be a map of "region = availability set ID"

```
  availability_set_id   = "${lookup(local.availavilityset_regions, "${var.region[count.index]}")}"
```

lookups the correct availability set ID for the vm region

### Actual Behavior

Code validates but Terraform crashes during plan

**Comment by juanjojulian** [comment]
Ok, I've found how to do this correctly:

```
availavilityset_regions = "${zipmap(azurerm_availability_set.main.*.location, azurerm_availability_set.main.*.id)}"
```

**Comment by apparentlymart** [comment]
Hi @juanjojulian! Sorry for this crash, and thanks for reporting it.

I think you've figured out here that this problem was caused by setting the key of a map to be a list instead of a string. Terraform was missing some validation to catch this and so it crashed when trying to evaluate the expression. This is the same root cause as in #18608, which is now fixed, so I'm going to close this one out too. The fix is in the master branch and will be included in the forthcoming v0.12.0 release.

I see you've already found a correct way to express what you needed here. There are some other examples in the other issue using newer features that will come in the forthcoming 0.12.0 release, but the `zipmap` approach you found will also continue working.

Thanks again for reporting this!
